OSDOCS#12855: Migrating the x86 control plane to arm64 architecture on AWS (#88421)
openshift/openshift-docs: modules/migrating-from-x86-to-arm-cp.adoc

master-001.example.com        Ready    control-plane,master   120d   v1.30.7   10.x.x.x    <none>        Red Hat Enterprise Linux CoreOS 4xx.xx.xxxxx-0   5.x.x-xxx.x.x.el9_xx.x86_64    cri-o://1.30.x
master-002.example.com        Ready    control-plane,master   120d   v1.30.7   10.x.x.x    <none>        Red Hat Enterprise Linux CoreOS 4xx.xx.xxxxx-0   5.x.x-xxx.x.x.el9_xx.x86_64    cri-o://1.30.x
master-003.example.com        Ready    control-plane,master   120d   v1.30.7   10.x.x.x    <none>        Red Hat Enterprise Linux CoreOS 4xx.xx.xxxxx-0   5.x.x-xxx.x.x.el9_xx.x86_64    cri-o://1.30.x
----
+
The `KERNEL-VERSION` field in the output indicates the architecture of the nodes.

. Check that your cluster uses the multi payload by running the following command:
+
[source,terminal]
----
$ oc adm release info -o jsonpath="{ .metadata.metadata}"
----
+
If you see the following output, the cluster is multi-architecture compatible.
+
[source,terminal]
----
{
 "release.openshift.io/architecture": "multi",
 "url": "https://access.redhat.com/errata/<errata_version>"
}
----
+

> If the cluster is not using the multi payload, migrate the cluster to a multi-architecture cluster. For more information, see "Migrating to a cluster with multi-architecture compute machines". 

Reviewer: is it better to use link for "Migrating to a cluster with multi-architecture compute machines" instead of text here?

Reviewer: aha, I saw you put all of the links together in "Additional resources", it's ok for me

. Update your image stream from single-architecture to multi-architecture by running the following command:
+
--
include::snippets/update-image-stream-to-multi-arch.adoc[]
--

. Get the `arm64` compatible Amazon Machine Image (AMI) for configuring the control plane machine set by running the following command:
+
[source,terminal]
----
$ oc get configmap/coreos-bootimages -n openshift-machine-config-operator -o jsonpath='{.data.stream}' | jq -r '.architectures.aarch64.images.aws.regions."<aws_region>".image' <1>
----

> <1> Replace `<aws_region>` with an AWS region. 

Reviewer: ```suggestion
<1> Replace `<aws_region>` with the AWS region where the current cluster is installed.   
```
we should get AMI for  current cluster installed region, there are many ways to get it, the below command is one way
`oc get infrastructure cluster -o jsonpath='{.status.platformStatus.aws.region}'`

Author: Updated as per the suggestion.

+
.Example output
[source,terminal]
----
ami-xxxxxxx
----

. Update the control plane machine set to support the `arm64` architecture by running the following command:
+
[source,terminal]
----
$ oc edit controlplanemachineset.machine.openshift.io cluster -n openshift-machine-api
----
+

> Update the `instanceType` to a type that supports the `arm64` architecture, and set the `ami` to an AMI that is compatible with the `arm64` architecture. For information about supported instance types, see "Tested instance types for AWS on 64-bit ARM infrastructures".

Author: Updated as per the suggestion.

+
For more information about configuring the control plane machine set for AWS, see "Control plane configuration options for {aws-full}".

.Verification

> . Verify that the control plane nodes are now running on the `arm64` architecture:

Reviewer: ```suggestion
. Verify that the control plane nodes are now running on the `arm64` architecture after the migration is complete:
```

Author: I think this additional info is redundant because this is part of verification (that is to be done only after completing the procedure)

// Module included in the following assemblies:
//
// * updating/updating_a_cluster/migrating-to-multi-payload.adoc

:_mod-docs-content-type: PROCEDURE
[id="migrating-from-x86-to-arm64-cp_{context}"]
= Migrating the x86 control plane to arm64 architecture on AWS

You can migrate the control plane in your cluster from `x86` to `arm64` architecture on {aws-full}.

.Prerequisites

> * You have access to a multi-architecture cluster.

Reviewer 2: To clarify, the cluster doesn't have to be multi-arch. It only needs to be running the multi payload.

Author: Thanks for the clarification!
Removed this pre-req as we have already talked about ensuring the cluster uses multi payload as part of the procedure.

* You have installed the {oc-first}.
* You logged in to `oc` as a user with `cluster-admin` privileges.

.Procedure

. Check the architecture of the control plane nodes by running the following command:
+
[source,terminal]
----
$ oc get nodes -o wide
----
+
.Example output
[source,terminal]
----
NAME                          STATUS   ROLES                  AGE    VERSION   INTERNAL-IP EXTERNAL-IP   OS-IMAGE                                         KERNEL-VERSION                 CONTAINER-RUNTIME

> worker-001.example.com        Ready    worker                 100d   v1.30.7   10.x.x.x    <none>        Red Hat Enterprise Linux CoreOS 4xx.xx.xxxxx-0   5.x.x-xxx.x.x.el9_xx.aarch64   cri-o://1.30.x

Reviewer 2: We should start with a cluster that only has x86 workers. We don't want to imply the the cluster has to have other architectures to start.

Author: Updated as per the suggestion.

+
[source,terminal]
----
$ oc edit controlplanemachineset.machine.openshift.io cluster -n openshift-machine-api
----
+
Update the `instanceType` to a type that supports the `arm64` architecture, and set the `ami` to an AMI that is compatible with the `arm64` architecture. For information about supported instance types, see "Tested instance types for AWS on 64-bit ARM infrastructures".
+
For more information about configuring the control plane machine set for AWS, see "Control plane configuration options for {aws-full}".

.Verification

. Verify that the control plane nodes are now running on the `arm64` architecture:
+
[source,terminal]
----
$ oc get nodes -o wide
----
+
.Example output
[source,terminal]
----
NAME                          STATUS   ROLES                  AGE    VERSION   INTERNAL-IP EXTERNAL-IP   OS-IMAGE                                         KERNEL-VERSION                 CONTAINER-RUNTIME
worker-001.example.com        Ready    worker                 100d   v1.30.7   10.x.x.x    <none>        Red Hat Enterprise Linux CoreOS 4xx.xx.xxxxx-0   5.x.x-xxx.x.x.el9_xx.aarch64   cri-o://1.30.x

> worker-002.example.com        Ready    worker                 98d    v1.30.7   10.x.x.x    <none>        Red Hat Enterprise Linux CoreOS 4xx.xx.xxxxx-0   5.x.x-xxx.x.x.el9_xx.aarch64   cri-o://1.30.x

Reviewer 2: same as above, lets keep all the workers as x86

Author: Updated as per the suggestion.

// Module included in the following assemblies:
//
// * updating/updating_a_cluster/migrating-to-multi-payload.adoc

:_mod-docs-content-type: PROCEDURE
[id="migrating-from-x86-to-arm64-cp_{context}"]
= Migrating the x86 control plane to arm64 architecture on AWS

You can migrate the control plane in your cluster from `x86` to `arm64` architecture on {aws-full}.

.Prerequisites

* You have access to a multi-architecture cluster.
* You have installed the {oc-first}.
* You logged in to `oc` as a user with `cluster-admin` privileges.

.Procedure

. Check the architecture of the control plane nodes by running the following command:
+
[source,terminal]
----
$ oc get nodes
----
+
.Example output
[source,terminal]
----
NAME                          STATUS   ROLES                  OS/Arch       AGE    VERSION  

> ip-10-0-17-211.ec2.internal   Ready    control-plane,master   linux/amd64  128m   v1.29.6+aba1e8d

Reviewer 2: `oc get nodes` doesn't output os/arch?
You can run `oc get nodes -o wide` and look at the `kernel-version` to determine the arch of a machine

Author: Updated as per the suggestion.

----

. Verify the release image architecture by running the following command:
+
[source,terminal]
----
$ oc adm release info
----
+
.Example output
+
[source,terminal]
----
Name:           4.16.5
Digest:         sha256:123abc456def789ghi012jkl345mno678pqr901stu234vwx567yz0
Created:        2024-07-25T09:48:57Z
OS/Arch:        linux/amd64
Manifests:      730
Metadata files: 1
Pull From: quay.io/openshift-release-dev/ocp-release@sha256:123abc456def789ghi012jkl345mno678pqr901stu234vwx567yz0
----
+

> The cluster must be running on the `x86` architecture.

Reviewer 2: We need to make sure the cluster has the 'multi' payload, if not they need to upgrade.
https://docs.openshift.com/container-platform/4.17/post_installation_configuration/configuring-multi-arch-compute-machines/creating-multi-arch-compute-nodes-aws.html#multi-architecture-verifying-cluster-compatibility_creating-multi-arch-compute-nodes-aws

Author: Updated as per the suggestion.

> . Verify that the import mode policy for the image streams is set to `PreserveOriginal` by running the following command:

Reviewer 2: The command to check and re-import images streams is here: https://docs.openshift.com/container-platform/4.17/openshift_images/image-streams-manage.html#images-imagestream-import-import-mode_image-streams-managing

Author: Updated as per the suggestion.

+
[source,terminal]
----
$ oc get configmap/coreos-bootimages -n openshift-machine-config-operator -o jsonpath='{.data.stream}' | jq -r '.architectures.aarch64.images.aws.regions."<aws_region>".image' <1>

> ----

Author: Updated as per the suggestion.

<1> Replace `<aws_region>` with an AWS region. 
//Any other replaceable values in the above command?
// What should be the output of this command? To add as an exmaple output.
+
.Example output
----
TBD
----

. Update the control plane machine set to support the `arm64` architecture by running the following command:
+
[source,terminal]
----
$ oc edit controlplanemachineset.machine.openshift.io cluster -n openshift-machine-api
----
// Any replaceable values in the above command?
// What changes should be done in the machine set after running this command? Which fields should be updated?
+

> Update the `instanceType` to a type that supports the `arm64` architecture, and set the `ami` to an AMI that is compatible with the `arm64` architecture.

Author: Updated as per the suggestion.

> <1> Replace `<aws_region>` with the AWS region where the current cluster is installed. You can get the AWS region for the installed cluster by running the following command:

Author: Updated as per the suggestion.

> $ oc get infrastructure cluster -o jsonpath='{.status.platformStatus.aws.region}

Reviewer: ```suggestion
$ oc get infrastructure cluster -o jsonpath='{.status.platformStatus.aws.region}'
```
missing a quotation mark.

----

> ami-xxxxxxx

Author: Fixed this issue.

----

. Update the control plane machine set to support the `arm64` architecture by running the following command:
+
[source,terminal]
----
$ oc edit controlplanemachineset.machine.openshift.io cluster -n openshift-machine-api
----
+

> Update the `instanceType` to a type that supports the `arm64` architecture, and set the `ami.id` to an AMI that is compatible with the `arm64` architecture. For information about supported instance types, see "Tested instance types for AWS on 64-bit ARM infrastructures".

Author: Updated as per the suggestion.
